cmd/config/apply.go

```diff
@@ -0,0 +1,154 @@
+		Example: configApplyExample,
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			cmdUtils.ExecuteParentHooks(cmd, args)
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return apply(cmd, options, configuration)
+		},
+	}
+	cmd.Flags().StringVarP(&options.configFile, "file", "f", "", "path to the configuration file")
+	err := cmd.MarkFlagRequired("file")
+	if err != nil {
+		return nil
+	}
+	return cmd
+}
+
+func apply(cmd *cobra.Command, options *configApplyOptions, configuration *config.Configuration) error {
+	payload := model.ConfigurationConfig{}
+	//in case this is running on a GitHub instance
+	gitWorkspace, present := os.LookupEnv("GITHUB_WORKSPACE")
+	_, isATest := os.LookupEnv("ARMORY_CLI_TEST")
+	if present && !isATest {
+		options.configFile = gitWorkspace + options.configFile
+	}
+	// read yaml file
+	file, err := ioutil.ReadFile(options.configFile)
+	if err != nil {
+		return fmt.Errorf("error trying to read the YAML file: %s", err)
+	}
+	cmd.SilenceUsage = true
+	// unmarshall data into struct
+	err = yaml.UnmarshalStrict(file, &payload)
+	if err != nil {
+		return fmt.Errorf("error invalid configuration object: %s", err)
+	}
+	configClient := configCmd.GetConfigClient(configuration)
+	if payload.Roles != nil {
+		err := processRoles(configClient, payload.Roles, cmd, payload.AllowAutoDelete)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func findDeletedRoles(rolesInConfigFile, existingRoles []model.RoleConfig) []string {
+	ma := make(map[string]bool, len(rolesInConfigFile))
+	var deletedRoles []string
+	for _, configRole := range rolesInConfigFile {
+		ma[configRole.Name] = true
+	}
+	for _, existinRole := range existingRoles {
+		if !ma[existinRole.Name] {
+			deletedRoles = append(deletedRoles, existinRole.Name)
+		}
+	}
+	return deletedRoles
+}
+
+func processRoles(configClient *configCmd.ConfigClient, rolesFromConfig []model.RoleConfig, cmd *cobra.Command, allowAutoDelete bool) error {
+	//get existing rolls
+	ctx, cancel := context.WithTimeout(configClient.ArmoryCloudClient.Context, time.Minute)
+	defer cancel()
+	// execute request
+	existingRoles, _, err := configClient.GetRoles(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting existing roles: %s", err)
+	}
+	//check to see if roll in config file exists already, if so perform a PUT, if not perform a POST to create
+	exists := false
+	for _, roleInConfig := range rolesFromConfig {
+		for _, roleInExisting := range existingRoles {
+			if roleInConfig.Name == roleInExisting.Name {
+				exists = true
+				//update existing role
+				ctx, cancel := context.WithTimeout(configClient.ArmoryCloudClient.Context, time.Minute)
+				defer cancel()
+				req, err := configCmd.UpdateRolesRequest(&roleInConfig)
+				_, _, err = configClient.UpdateRole(ctx, req)
+				if err != nil {
+					return fmt.Errorf("error trying to update role: %s", err)
+				}
+				_, err = fmt.Fprintln(cmd.OutOrStdout(), "Updated role: "+roleInConfig.Name)
```

Comment:
Is there a reason to write to stdout directly instead of using a logger? It looks like we're using a logger in other places in the CLI.

Reply:
This seems to be inconsistent across the entire project. I filed a tech-debt ticket for us to go through and clean this up (and ensure we are using best practices). https://armory.atlassian.net/browse/CDAAS-1631